x-pack/plugins/code/server/ctags/ctags_reader.ts

```diff
@@ -0,0 +1,147 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { Tag, TagFields } from '../../model';
+import { LoggerFactory } from '../utils/log_factory';
+
+export class CtagsReader {
+  private tags: Tag[] = [];
+
+  private MIN_METHOD_LINE_LENGTH = 6;
+  private MAX_METHOD_LINE_LENGTH = 1030;
+
+  constructor(
+    readonly loggerFactory: LoggerFactory
+  ) {}
+
+  public getTags(): Tag[] {
+    return this.tags;
+  }
+
+  public readLine(tagLine: string) {
+    // const logger = this.loggerFactory.getLogger(['code', 'ctagsReader']);
+
+    if (tagLine === null) {
+      return;
+    }
+
+    let p = tagLine.indexOf('\t');
+    if (p <= 0) {
+      return;
+    }
+    const def = tagLine.substring(0, p);
+    const mstart = tagLine.indexOf('\t', p + 1);
+
+    let kind = null;
+
+    let lp = tagLine.length;
+    let tagFieldMap = new Map();
+    while((p = tagLine.lastIndexOf('\t', lp - 1)) > 0) {
+      const fld = tagLine.substring(p + 1, lp);
+      lp = p;
+
+      const sep = fld.indexOf(':');
+      let matchTagField = null;
+      if (sep != -1) {
+        for (let tagField in TagFields) {
+          if (TagFields[tagField].charAt(0) !== fld.charAt(0)) {
+            continue;
```

Comment:
IMHO, we convert the `for loop` to the bellow pattern to trim the `continue` statement.
```ts
for (let tagField in TagFields) {
  if (TagFields[tagField].charAt(0) == fld.charAt(0)) {
    matchTagField = TagFields[tagField];
  }
}
```

Reply:
Here the comparison between `TagFields[tagField].charAt(0)` and `TagFields[tagField].charAt(0)` is used to accelerate string matching, since there's only three different types of fields now (line, signature, class) starting with different letters, which means we can compare the first letter to quickly refuse a comparison. Semantically, this is a process of pruning. So I think it's a normal **pre-pruning** coding style.

Reply:
Totally make sense, I neglect the code meaning!